node/src/reactor/validator.rs

```diff
@@ -783,36 +782,51 @@ impl reactor::Reactor for Reactor {
+                                        block_height, sender, error
+                                    );
+                                    None
+                                });
+                            let chainspec = Arc::clone(self.chainspec_loader.chainspec());
 
-                            match Message::new_get_response(
-                                &fetched_or_not_found_block_header_and_finality_signatures,
-                            ) {
-                                Ok(message) => {
-                                    return effect_builder.send_message(sender, message).ignore();
+                            return async move {
+                                let mut fully_signed = false;
+                                if let Some(header_with_sigs) = &maybe_header_with_sigs {
+                                    if let Some(validator_weights) = effect_builder
+                                        .get_era_validators(header_with_sigs.block_header.era_id())
+                                        .await
+                                    {
+                                        fully_signed =
+                                            linear_chain_sync::weigh_finality_signatures(
+                                                &validator_weights,
+                                                chainspec
+                                                    .highway_config
+                                                    .finality_threshold_fraction,
+                                                &header_with_sigs.block_signatures,
+                                            )
+                                            .is_ok();
+                                    }
                                 }
-                                Err(error) => {
-                                    error!("failed to create get-response: {}", error);
-                                    return Effects::new();
+                                let fetch_or_not_found_header_with_sigs =
+                                    match maybe_header_with_sigs.filter(|_| fully_signed) {
+                                        None => FetchedOrNotFound::NotFound(block_height),
+                                        Some(header) => FetchedOrNotFound::Fetched(header),
+                                    };
+                                match Message::new_get_response(
+                                    &fetch_or_not_found_header_with_sigs,
+                                ) {
+                                    Ok(message) => {
+                                        effect_builder.send_message(sender, message).await
+                                    }
+                                    Err(error) => error!("failed to create get-response {}", error),
                                 }
-                            };
+                            }
```

Comment:
This looks very similar to the method in `linear_chain_sync::operations`, is that what the `TODO` there is about?

Reply:
It looks like I'll probably have to rewrite this PR anyway after `feat-fast-sync` has been rebased and @xcthulhu's latest refactorings merged. Then I'll try to deduplicate this, too.